Base implementation for TsAssets

kenju opened the pull request:
Base implementation for TsAssets. 

`rake ts_assets:generate` で、`assets.tsx` が生成されます。
テストも書いています。

## ディレクトリ構成

```
lib/
  ts_assets.rb
  ts_assets/
    application_generator.rb
    generators/
      const_generator.rb
      react_generator.rb
    models/
      asset_meta_info.rb
      content.rb
```

## `assets.tsx` 生成までの処理の流れ

コールグラフで説明した場合（コアとなるメソッドのみに絞り、それ以外は省略して書いています）：

- `TsAssets.generate`（ライブラリを使う側はこのクラスメソッドを適切なオプションとともにコールするだけ）
  - `ApplicationGenerator#build_mapping`（各アセットファイルへのパスをkey、`AssetMetaInfo`をvalueにしたHasyを返す）
    - `AssetMetaInfo#initialize`（各Generatorsがファイル内容を組み立てる際に必要なアセット情報をまるっとまとめたModel。画像width/heightやdigest_pathをインスタンス生成時に取得する）
  - `ApplicationGenerator#generate`（各Generatorsを呼び出し、適切なファイル形式に整形して返す）
    - `ConstGenerator#generate`（`PATH_***`形式のパスを返す）
    - `ReactGenerator#generate`
      - `ReactGenerator#merge_mapping_with_same_descriptors`（`srcSet`を組み立てる際に、`*@1x`, `*@2x`の接尾辞がついたファイル名を同じものとして扱いやすくするため、`ApplicationGenerator#build_mapping`が構築したマッピングを元に再構築する。ここでは、keyはそのまま、valueが`Array<AssetMetaInfo>`に）

kenju committed:
update gemspec TODO

kenju committed:
add spec/ts_assets/application_generator_spec.rb spec/ts_assets/generators spec/ts_assets/models and spec/spec_helper

kenju committed:
add lib/ts_assets lib/ts_assets.rb

kenju committed:
git ignore .idea

kenju committed:
add fastimage gem

kenju committed:
introduce all spec

kenju committed:
update README

kenju committed:
updaet gemspec to include run_time_dependencies

kenju commented on Rakefile:
```diff
@@ -1,6 +1,13 @@
 
 RSpec::Core::RakeTask.new(:spec)
 
+task :generate do
```
ライブラリを使う側は、このような `rake task` を定義して使うことになります。

gfx commented on lib/ts_assets/generators/react_generator.rb:
```diff
@@ -0,0 +1,83 @@
+
+        <<~TS
+          /** #{path} */
+          export function #{component_name}(props: any) {
```
`props: any`  ではなくちゃんとインターフェイスを定義してください。

kenju replied:
https://github.com/bitjourney/ts_assets-rails/pull/1/commits/532e25369e6798216559edcb088da0dd5b655ebb

`React.HTMLProps<HTMLImageElement>` 使うようにしました！
（参考：[lib.d.ts で定義されている HTMLImageElement](https://github.com/Microsoft/TypeScript/blob/master/lib/lib.d.ts#L9406)）

gfx commented on lib/ts_assets/generators/react_generator.rb:
```diff
@@ -0,0 +1,83 @@
+          /** #{path} */
+          export function #{component_name}(props: any) {
+              return <img alt="#{alt}"
+                          width={#{width}}
```
widthだけではなくheightもあったほうがいいと思いますが、なぜwidthだけなんですか？

kenju replied:
よく考えたんですが、`height` は指定した方が良さそうですね。
なので、`height` つけます！

---

補足。

- CSSで `width`のみを指定してレイアウトを調整している `<img/>`
- かつ、その指定している `width` が、本来の画像ファイルの `width` よりも小さい

場合に、「CSSで `height` も指定する必要性が使用者側に生じる」ことになるのがイマイチかなと思ったわけです。例えば、

- (width, height) = (640px, 320px)
- CSSで `width=100%` と指定（親のElementの `width=422px` なので、実質的には `width < 422px`

している箇所で、`<img height="320" ... />` を組み込むようにすると、画像が間延びしてしまいます。

![bitjourney3 kibela 2017-07-04 12-45-53-chm85](https://user-images.githubusercontent.com/4186381/27814450-6a8bbf30-60b7-11e7-89c8-58d0fb0e0362.png)

最初、ライブラリを導入することでCSSを書き換える必要が新たに生じることは何かおかしいのでは？と思ったわけですが、そもそも画像のプレロード時前の幅・高さを確保するために`height`を指定すべきで、このような場合ではCSSで新たに `height` を指定するのが正しい、と今は理解しています。

kenju replied:
つけました！

https://github.com/bitjourney/ts_assets-rails/pull/1/commits/c9e23c8ac1326689e1a2aa154122dd17029439ac

kenju committed:
update YARD

kenju committed:
declare props as React.HTMLProps<HTMLImageElement>

kenju committed:
add height attribute to img tag

kenju commented on lib/ts_assets/models/asset_meta_info.rb:
```diff
@@ -0,0 +1,86 @@
+
+      # @return [String]
+      def digest_path
+        environment.find_asset(asset_path).digest_path
```
`Rails.application.asset.find_asset(...)` と基本的には同じです。

なお、この `asset_path` には、[SprocketsにおけるLogical Path](https://github.com/rails/sprockets#logical-paths)を渡さなくてはいけません。具体的にいうと、`app/asssets/images/webhook/slack.png` という画像があった時

- 🆗  "webhook/slack.png"
- 🆖  "app/assets/images/webhook/slack.png"
- 🆖  "images/webhook/slack.png"
- 🆖  "slack.png"

ということです。

kenju commented on lib/ts_assets/models/asset_meta_info.rb:
```diff
@@ -0,0 +1,86 @@
+
+        @width, @height = FastImage.size(full_path)
+
+        if has_descriptor?
```
画像名が `slack_icon@1x` のように、`*@1x` などのPixel Descriptorの接尾辞が存在する時、

- `@asset_path_without_descriptor` : `slack_icon`
- `@descriptor`: `1x`

を保持します。`@asset_path_without_descriptor`を持たせているのは、

-  `slack_icon@1x`
-  `slack_icon@2x`

という違うPixel Descriptor値の別画像アセットが存在する時、[ReactGenerator](https://github.com/bitjourney/ts_assets-rails/pull/1/files#diff-2310d072461689fcfae4478bb475acb1R27) において同一の画像アセットとして扱えるようにするためです。

kenju commented on spec/build/assets.tsx:
```diff
@@ -0,0 +1,35 @@
+/* tslint:disable */
```
これはテストで比較するためのスタブです。`rake generate` の生成結果です。

基本的に画像アセットが変化しない場合Digestも変化しないので、IntegrationTestの意味合いで、トップレベルで `TsAssets.generate` を読んだ時にこの生成ファイルが常に不変であることを[ここのテスト](https://github.com/bitjourney/ts_assets-rails/pull/1/files#diff-e90fda9463de91b6e4f6d8d472781f0eR9)保証しています。

gfx replied:
コード生成系を初めて書くときはやりがちなんですが、生成したコードが一字一句golden resultと同じかどうかというテストは、すべきでないです。ちょっと修正するたびにgolden resultのほうを修正しなければならずメンテナンスが非常に煩雑になるわりに、本質的には意味がないからです。特にパッチを送ろうとする第三者からするとこの種のテストは無視しづらく、無駄にハードルが上がります。

テストをしたいなら生成されたコードを使うテストを書きましょう。

kenju replied:
> ちょっと修正するたびにgolden resultのほうを修正しなければならずメンテナンスが非常に煩雑になる

たしかに。。
「アセットが変わらなければ生成ファイルが常に不変であることを保証できる」というのは、「テストが変わるたびにdiffを目視で確認しながらそれが意図する変更であることを確認する」メンテナンス作業があってこそのものですね。

テストのbefore/afterで毎回`build/assets.tsx` を生成・破棄する形にしたのですが、どうでしょうか？

https://github.com/bitjourney/ts_assets-rails/pull/1/commits/680ee6353cc08919dbcba547c8eaee11f488bd29

gfx replied:
いちいち破棄すると今度は目視で確認しにくくなるので、たんに .gitignore で無視するのがいいと思います。

kenju replied:
> いちいち破棄すると今度は目視で確認しにくくなるので

あーそうか 🙇 
そうしました
https://github.com/bitjourney/ts_assets-rails/pull/1/commits/2929dc40891798452221c7e139bd6f583dd333fa

kenju commented on lib/ts_assets/models/asset_meta_info.rb:
```diff
@@ -0,0 +1,86 @@
+
+      # @return [String]
+      def asset_path
+        full_path.gsub(%r{^#{include_path}/}, '')
```
https://github.com/bitjourney/ts_assets-rails/pull/1/files#r125386206 に書いた通り、Logical Pathが必要なので、`full_path` から `include_path` を削除するためのロジックです。

kenju commented on lib/ts_assets/generators/react_generator.rb:
```diff
@@ -0,0 +1,88 @@
+
+      # @param [Hash<String, TsAssets::Models::AssetMetaInfo>] mapping
+      # @return [Hash<String, [TsAssets::Models::AssetMetaInfo]>]
+      def merge_mapping_with_same_descriptors(mapping)
```
[テスト](https://github.com/bitjourney/ts_assets-rails/pull/1/files#diff-b7f29c830e32caab89ad0f004f17d3f9R49) を見ていただくとイメージがつけめると思うのですが、ここでは `***@1x`, `***@2x` などの異なるPixel Descriptorを持つアセットを丸っと `srcSet` にいい感じにセットするために、`ApplicationGenerator#build_mapping` から渡されたマッピングを再構築しています。

具体例）

```rb
{"spec/assets/images/svg/ruby-icon.svg"=>
  #<TsAssets::Models::AssetMetaInfo:0x007f8ca37cc838
   @asset_path_without_descriptor="svg/ruby-icon",
   @descriptor="1x",
   @full_path="spec/assets/images/svg/ruby-icon.svg",
   @include_path="spec/assets/images">,

 "spec/assets/images/webhook/slack_icon@1x.png"=>
  #<TsAssets::Models::AssetMetaInfo:0x007f8ca37c5358
   @asset_path_without_descriptor="webhook/slack_icon",
   @descriptor="1x",
   @full_path="spec/assets/images/webhook/slack_icon@1x.png",
   @include_path="spec/assets/images">,

 "spec/assets/images/webhook/slack_icon@2x.png"=>
  #<TsAssets::Models::AssetMetaInfo:0x007f8ca37c44f8
   @asset_path_without_descriptor="webhook/slack_icon",
   @descriptor="2x",
   @full_path="spec/assets/images/webhook/slack_icon@2x.png",
   @include_path="spec/assets/images">
}
```

のようなマッピングがあった時に、

```rb
{"svg/ruby-icon"=>
  [#<TsAssets::Models::AssetMetaInfo:0x007f8ca37cc838
    @asset_path_without_descriptor="svg/ruby-icon",
    @descriptor="1x",
    @full_path="spec/assets/images/svg/ruby-icon.svg",
    @include_path="spec/assets/images">],

 "webhook/slack_icon"=>
  [#<TsAssets::Models::AssetMetaInfo:0x007f8ca37c5358
    @asset_path_without_descriptor="webhook/slack_icon",
    @descriptor="1x",
    @full_path="spec/assets/images/webhook/slack_icon@1x.png",
    @include_path="spec/assets/images">,
   #<TsAssets::Models::AssetMetaInfo:0x007f8ca37c44f8
    @asset_path_without_descriptor="webhook/slack_icon",
    @descriptor="2x",
    @full_path="spec/assets/images/webhook/slack_icon@2x.png",
    @include_path="spec/assets/images">]}
```

というように、同じ画像だけど異なるPixel Descriptor持つ場合`Array<PixelDescriptor>` に格納する形に再構築しています。

kenju commented on lib/ts_assets/application_generator.rb:
```diff
@@ -0,0 +1,61 @@
+
+    # @param [String] include_path
+    # @return [Hash]
+    def build_mapping(include_path)
```
Descriptionに書いた全体のコールグラフのように、

- ApplicationGenerator#build_mapping で生成したマッピングを
- ConstGenerator, ReactGenerator それぞれに渡して使ってもらう

という流れです。

kenju commented on ts_assets.gemspec:
```diff
@@ -29,6 +29,11 @@ Gem::Specification.new do |spec|
   spec.require_paths = ["lib"]
 
+  spec.add_runtime_dependency "railties", "~> 5.0"
+  spec.add_runtime_dependency 'fastimage'
```
画像の幅・高さ取得には `fastimage` gemを使っています。

kenju commented on lib/ts_assets/application_generator.rb:
```diff
@@ -0,0 +1,61 @@
+    def initialize(include:)
+      @mapping = build_mapping(include)
+
+      environment.append_path(include)
```
https://github.com/rails/sprockets#manipulating-the-load-path

`append_path` にパスを与えることによって、Sprockets::Environment が画像をロードする際に探しに行くパスを追加できます。

kenju commented:
@bitjourney/kibela-developers レビューよろしくお願いします！ 🙏

gfx commented on spec/ts_assets/generators/react_generator_spec.rb:
```diff
@@ -0,0 +1,104 @@
+
+  describe '#build_component_name' do
+    it do
+      expect(generator.send(:build_component_name, "svg/ruby-icon")).to eq "ImageSvgRubyIcon"
```
private methodを `send` でtestするのはやめてください。private methodは実装の詳細で、普通はテストの対象ではありません。

どうしてもtestしたければpublic methodにしてください。しかしこのgemの場合は生成されたJSを走らせる結合テストがあれば十分だと思います。

kenju replied:
> private methodを send でtestするのはやめてください。private methodは実装の詳細で、普通はテストの対象ではありません。

private method を細かくテストしても意味ないし、メンテナンスコストも上がってしまいますね。それに`ts_asseets_spec.rb`だけテストすればファイル生成に関しては十分テストできてそうです。

ということで、Generators周りのテストを丸っと消しました！
https://github.com/bitjourney/ts_assets-rails/pull/1/commits/b44956d78b21789b7a56e3195c894054e32fe2e2

gfx replied:
細かくテストするのが意味がないとは言ってませんよ、念のため。

private methodを細かくテストしなければならないとしたら設計ミスの可能性が高いと思ってます。テストが必要なのであれば、別モジュール（ないしクラス）にしてpublic methodをテストすればいいのです。そのモジュール（ないしクラス）のpublicインターフェイスを介してテストできる範囲のみテストするように全体の設計をしてください。

kenju replied:
> private methodを細かくテストしなければならないとしたら設計ミスの可能性が高いと思ってます。

= つまり、「実装の詳細にテストしたいことが詰め込みすぎている」= 「そのクラスがたくさんの責任・ロジックを担っている」= 「設計ミス」という感じですかね 📝

kenju committed:
remove redundant tests

because
- testing private method is usually avoided since it is implementation details
- testing ts_assets_spec for generation is enought as an integration test

kenju committed:
git rm spec/build/assets.tsx and generate assets.tsx everytime for testing

gfx commented on lib/ts_assets/application_generator.rb:
```diff
@@ -0,0 +1,61 @@
+    def build_mapping(include_path)
+      mapping = {}
+
+      Dir.glob("#{include_path}/**/*.*").map do |full_path|
```
戻り値を使わない場合、 `map` ではなく `each` を使いましょう。

kenju replied:
そうでした、ここは `map`を直接返却しなくなったので `each` に修正しました

https://github.com/bitjourney/ts_assets-rails/pull/1/commits/c64dcfc86caa8c77a7e1406bd8ce6974826b4d17

gfx commented on lib/ts_assets/generators/react_generator.rb:
```diff
@@ -0,0 +1,88 @@
+      # @param [String] asset_path
+      # @return [String]
+      def build_component_name(asset_path)
+        invalid_chars = %r{[@_/-]}
```
ファイル名には `+` とかも普通に使いますよ。こういう場合、ブラックリストにするのが間違いのもとです。

`/[a-zA-Z]0-9_]/` というホワイトリストにするのがいいと思います。

gfx commented on lib/ts_assets/generators/react_generator.rb:
```diff
@@ -0,0 +1,88 @@
+      # @return [String]
+      def build_component_name(asset_path)
+        invalid_chars = %r{[@_/-]}
+        normalised_name = asset_path.split(invalid_chars).map(&:capitalize).join
```
`name.gsub(/[^a-zA-Z0-9_]+/, '_').camelize` とかがいいんじゃないですかね。

kenju replied:
[`camelize`](https://github.com/rails/rails/blob/24bacb3e956e36414847cb250bbd6c6d9c6cadd7/activesupport/lib/active_support/core_ext/string/inflections.rb#L89) があったのですね！

正規表現のホワイトリスト化の件も修正しました！ 📝 

https://github.com/bitjourney/ts_assets-rails/pull/1/commits/8238b54744651e953f690d5c1780e6d913d277f3

gfx commented on lib/ts_assets/models/asset_meta_info.rb:
```diff
@@ -0,0 +1,86 @@
+
+      # @return [String]
+      def normalised_path
+        invalid_chars = %r{[/.@-]}
```
ここもホワイトリストがいいですね。

kenju committed:
s/map/each for Dir.glob

kenju committed:
fix regex for normalising from black-listing to white-listing

kenju commented:
~~諸々修正しました！~~
~~お時間ある時改めてレビューお願いします 🙏~~

JS側のIntergration Testを書きます！

kenju committed:
add integration test with tslint

kenju commented:
https://github.com/bitjourney/ts_assets-rails/pull/1/commits/4403be3a1b5f3ad4f45f6273088d883655edcf8f#diff-2e0ee9ff1d3c789f5547e2723e87fbd2

JSでのIntegration Test追加しました！TSLintも追加しました！

gfx commented:
LGTM 👏

michiomochi commented on package.json:
```diff
@@ -0,0 +1,52 @@
+{
+  "name": "ts_assets-rails",
+  "private": true,
```
[nits] privateは必要なさそう。

gfx replied:
リリースするものではないのでprivateが必要なのです。

michiomochi replied:
ああ、そうか！このレポジトリの主体はgemか。完全に的外れだった 🙇

michiomochi commented on package.json:
```diff
@@ -0,0 +1,52 @@
+    "url": "git+ssh://git@github.com/bitjourney/ts_assets-rails.git"
+  },
+  "author": "kenju",
+  "license": "Apache-2.0",
```
licenceってApache-2.0でいきます？
あとLICENCE.txt配置したほうがよさそう。

gfx replied:
licenseはgemと同じものにしてください。どちらもapache 2.0にするのがいいかな。

kenju replied:
Apache-2.0を追加しましたー 🙏 

https://github.com/bitjourney/ts_assets-rails/pull/1/commits/3810d8e384cf086d45e640f6e8a6a9773e086668

kenju committed:
add license with Apache-2.0

michiomochi commented on .circleci/config.yml:
```diff
@@ -0,0 +1,51 @@
+jobs:
+  build:
+    docker:      
+      - image: bitjourney/kibela-ci:latest
```
rubyのofficial image使ってビルドするようにしたほうがいいかなと思います。
kibela-ciはビルドのされ方がブラックボックス(僕達は見れるが)なのでpublic repoに使うのは不適当かなと。

kenju replied:
たしかにですね 🙏 `ruby:latest` を Primary Imageに使うように変更しました！

https://github.com/bitjourney/ts_assets-rails/pull/1/commits/3b6a484dba91c4f87fd77307fd955c3dcef30e18

michiomochi commented on .circleci/config.yml:
```diff
@@ -0,0 +1,51 @@
+    environment:
+      BUNDLE_DIR: vendor/bundle
+    
+    working_directory: /home/ubuntu/kibela
```
[nits] kibela関係ないのでkibelaじゃないほうがよさそう。

kenju committed:
use official ruby docker image as a primary container

kenju committed:
install react-test-renderer because it is required by enzyme

michiomochi commented:
CircleCIでRubyのRestore cacheがコケてそう。
Gemfile.lockはコミットしてないからないよといわれてるんだな。
![continuous integration and deployment 2017-07-06 17-08-17](https://user-images.githubusercontent.com/1830471/27901631-c042d442-626d-11e7-9083-acb26f940796.png)

michiomochi commented:
いや、gemレポジトリではGemfile.lockはコミットしてはだめなのでは。
CircleCIの設定を見直すべきのように思います。

kenju commented:
依存ライブラリのバージョン固定してしまうから、gemは[普通に入れちゃだめ](http://yehudakatz.com/2010/12/16/clarifying-the-roles-of-the-gemspec-and-gemfile/)ですね。Circle CIの設定考えてみます！

kenju committed:
change circleci cache key for gems

kenju committed:
add bundle exec when generating assets in preset of npm scripts

kenju commented:
> Gemfile.lockはコミットしてないからないよといわれてるんだな。

`checksum Gemfile.lock` の代わりに `checksum ts_assets.gemspec` をキャッシュキーとして見るようにしました -> https://github.com/bitjourney/ts_assets-rails/pull/1/commits/d1d580c264e836b88e084171fb43815ee398bd17

michiomochi commented:
LGTMです！ 👏 🆒

kenju commented:
レビューありがとうございました！！
